### PR TITLE
Allow non CS entropy sources for `psa_generate_random()`

### PR DESCRIPTION
**Pull request by valeriosetti**
## Workflow of this PR

- starts with some renaming of entropy kconfigs in order to have them more aligned to Mbed TLS build symbols (which simplifies understanding their meaning/effect).
- improves the description of PSA kconfigs related to RNGs.
- align entropy gathering between legacy `mbedtls_hardware_poll()` and PSA `mbedtls_psa_external_get_random()`.
- and finally it adds a new kconfig `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CS` to allow non-CS entropy sources also for PSA
    - this is meant to be used only for tests (ex: testing a psa_get_random()` also on platforms that do not have CS sources), but it must be disabled in production. A warning is raised whenever this feature is enabled.

## Improvements from this PR

Prior to this PR when Mbed TLS's PSA crypto was to be built:

- `MBEDTLS_PSA_CRYPTO_LEGACY_RNG` was to be used on platforms that do not have an entropy driver
    - this required also `MBEDTLS_ENTROPY_ENABLED` to add a Zephyr's entropy callback from Mbed TLS's entropy module
- `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` was to be used on platforms that have an entropy driver.

After this PR it will be possible to enable `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` + `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG` on all platforms. The latter symbols must be disabled in production of course, but for CI purposes it allows a more uniform usage of Mbed TLS's PSA crypto across the Zephyr's tessts and samples.

### CODE REVIEW

File: modules/mbedtls/Kconfig.tls-generic [hunk 1]
```diff
@@ -387,12 +387,22 @@ config MBEDTLS_HAVE_ASM
 config MBEDTLS_ENTROPY_C
 	bool "Mbed TLS entropy accumulator"
 	depends on MBEDTLS_SHA256 || MBEDTLS_SHA384 || MBEDTLS_SHA512
-	default y if MBEDTLS_ZEPHYR_ENTROPY
 	help
 	  This module gathers entropy data from enabled entropy sources. It's
 	  mostly used in conjunction with CTR_DRBG or HMAC_DRBG to create
 	  a deterministic random number generator.
 
+config MBEDTLS_ZEPHYR_ENTROPY_POLL
```
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:395]: Thinking about the naming of this... Would `MBEDTLS_POLL_ZEPHYR_ENTROPY` (or `MBEDTLS_ENTROPY_POLL_ZEPHYR` express more clearly that `Mbed TLS polls Zephyr for entropy`?

File: tests/modules/uoscore/prj.conf [hunk 2]
```diff
@@ -12,7 +12,7 @@ CONFIG_UOSCORE=y
 CONFIG_MBEDTLS=y
 CONFIG_MBEDTLS_ENABLE_HEAP=y
 CONFIG_MBEDTLS_HEAP_SIZE=2048
-CONFIG_MBEDTLS_ZEPHYR_ENTROPY=y
+CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL=y
```
Comment (tomi-font) [inline tests/modules/uoscore/prj.conf:15]:
```suggestion
CONFIG_MBEDTLS_ENTROPY_C=y
CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL=y
```

File: modules/mbedtls/Kconfig.tls-generic [hunk 3]
```diff
@@ -483,16 +483,33 @@ config MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
...
 	help
-	  Use cryptographically secure random generator to provide random data
-	  instead of legacy MbedTLS modules (ENTROPY + CTR_DRBG/HMAC_DRBG).
+	  Use a cryptographically secure random generator to provide random data
+	  instead of legacy Mbed TLS modules. This has a smaller footprint
+	  than its legacy counterpart MBEDTLS_PSA_CRYPTO_LEGACY_RNG, but it
+	  requires a cryptographically secure random number generator (CSPRNG)
+	  to be available in the system. If no CSPRNG is available:
+	  - there is no fallback to weak entropy random generators.
+	  - the PSA crypto initialization will fail and none of the PSA APIs
```
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:491]: Thinking out loud: could we find out about this at build time to avoid failing only at runtime?

File: modules/mbedtls/Kconfig.tls-generic [hunk 4]
```diff
@@ -467,21 +481,49 @@ choice MBEDTLS_PSA_CRYPTO_RND_SOURCE
...
+	    will not be available.
```
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:492]:
```suggestion
	  - the initialization of PSA Crypto will fail and none of its API
	    will be available.
```

File: modules/mbedtls/Kconfig.tls-generic [hunk 5]
```diff
@@ -467,21 +481,49 @@ choice MBEDTLS_PSA_CRYPTO_RND_SOURCE
...
-	  Use legacy MbedTLS modules (ENTROPY + CTR_DRBG/HMAC_DRBG) as random
-	  source generators.
+	  Use legacy Mbed TLS modules to generate random data. In this
+	  configuration the entropy module is used to gather some data and then
+	  either ctr_drbg or hmac_drbg are applied on top of it to improve
+	  the randomness.
+	  Security level in this case really depends on the type of entropy
+	  sources which are enabled in the system: if weak entropy sources are
+	  used, then the generated data will only be pseudo random. Strong
+	  entropy sources are really suggested (if possible) to have real
```
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:506]:
```suggestion
	  entropy sources are strongly recommended (if possible) to have real
```

File: modules/mbedtls/Kconfig.tls-generic [hunk 6]
```diff
@@ -467,21 +481,49 @@ choice MBEDTLS_PSA_CRYPTO_RND_SOURCE
...
+	  random data.
+	  The main difference between this implementation and the
+	  MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG counterpart is just in terms of
+	  code size: this brings in more legacy modules and therefore it has
```
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:510]: Really? Aren't there other significant differences? (Some of them are stated in you new description of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`.)

File: modules/mbedtls/CMakeLists.txt [hunk 7]
```diff
@@ -11,6 +11,18 @@ zephyr_interface_library_named(mbedTLS)
...
+    if(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG)
+      message(WARNING "
+        Warning: non cryptographycally secure sources are enabled for psa_generate_random().
+        This is meant to be used only for tests, not in production!")
+    else()
+      if(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG AND NOT CONFIG_CSPRNG_ENABLED)
+        message(WARNING "
+          MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG is set but there is
+          no CSPRNG enabled.")
+      endif()
```
Comment (tomi-font) [inline modules/mbedtls/CMakeLists.txt:23]: Couldn't we just make the build fail in this case? If `EXTERNAL_RNG` is enabled, CSPRNG is not enabled, nor is `RNG_ALLOW_NON_CSPRNG`, then `psa_crypto_init()` will always fail right?

File: modules/mbedtls/CMakeLists.txt [hunk 8]
```diff
@@ -11,6 +11,18 @@ zephyr_interface_library_named(mbedTLS)
       message(WARNING "No entropy device on the system, using fake entropy source!")
     endif()
 
+    if(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG)
+      message(WARNING "
+        Warning: non cryptographycally secure sources are enabled for psa_generate_random().
```
Comment (tomi-font) [inline modules/mbedtls/CMakeLists.txt:16]: I think no need for `Warning:` here because CMake will already indicate that.

File: modules/mbedtls/Kconfig.tls-generic [hunk 9]
```diff
@@ -513,6 +512,18 @@ config MBEDTLS_PSA_CRYPTO_LEGACY_RNG
 
 endchoice
 
+config MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG
```
Comment (tomi-font) [inline modules/mbedtls/Kconfig.tls-generic:515]: Not sure about this suggestion, but this new Kconfig option could potentially be added as part of the `choice`, removing the dependency with `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` (and putting put the dependency of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` on `CSPRNG_ENABLED`)

File: modules/mbedtls/zephyr_init.c
Comment (tomi-font) [inline modules/mbedtls/zephyr_init.c:1]: Not introduced here, but this file has more than just initialization... Could rename/split it?

File: modules/mbedtls/zephyr_init.c [hunk 10]
```diff
@@ -134,13 +141,17 @@ psa_status_t mbedtls_psa_external_get_random(
 	uint8_t *output, size_t output_size, size_t *output_length)
 {
 	(void) context;
+	int ret;
 
-	if (sys_csrand_get(output, output_size) != 0) {
+	ret = get_random_data(output, output_size,
+		IS_ENABLED(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG) ?
+		false : true);
```
Comment (tomi-font) [inline modules/mbedtls/zephyr_init.c:148]: Suggest simplifying this by removing the ternary operator and turning `cs_only` in `get_random_data()` into `allow_non_cs`:
```suggestion
		IS_ENABLED(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG));
```

File: modules/mbedtls/zephyr_init.c [hunk 11]
```diff
@@ -122,6 +84,51 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
 	return (mbedtls_ms_time_t)k_uptime_get();
 }
 
+#if defined(CONFIG_MBEDTLS_ZEPHYR_ENTROPY_POLL) || defined(CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+static int get_random_data(uint8_t *output, size_t output_size, bool cs_only)
+{
+	int ret = MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
```
Comment (tomi-font) [inline modules/mbedtls/zephyr_init.c:90]:
```suggestion
	int ret = MBEDTLS_ERR_ENTROPY_NO_SOURCES_DEFINED;
```

File: modules/mbedtls/zephyr_init.c [hunk 12]
```diff
@@ -122,6 +84,51 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
...
+	int ret;
+	uint16_t request_len = len > UINT16_MAX ? UINT16_MAX : len;
+
+	ARG_UNUSED(data);
+
+	if (output == NULL || olen == NULL || len == 0) {
+		return MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+	}
+
+	ret = get_random_data(output, len, false);
```
Comment (tomi-font) [inline modules/mbedtls/zephyr_init.c:121]: Noting that here (in `mbedtls_hardware_poll()`) we're completely dropping the use of `CONFIG_ENTROPY_HAS_DRIVER`/`entropy_get_entropy()` (replacing it by `CONFIG_CSPRNG_ENABLED`/`sys_csrand_get()`).
Is this intended and if so why?

File: tests/crypto/mbedtls_psa/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(mbedtls)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
```
Comment (tomi-font) [inline tests/crypto/mbedtls_psa/CMakeLists.txt:10]: Not sure globbing is allowed. You may as well replace this by just
```suggestion
target_sources(app PRIVATE src/main.c)
```

File: tests/crypto/mbedtls_psa/testcase.yaml [hunk 14]
```diff
@@ -0,0 +1,37 @@
+# The goal here is to showcase that Mbed TLS's PSA implementation can be
```
Comment (tomi-font) [inline tests/crypto/mbedtls_psa/testcase.yaml:1]: I think it could help to make sure that we talk about things naming them properly, i.e. here PSA Crypto instead of just PSA, in order to avoid spreading more confusion to the world.
```suggestion
# The goal here is to showcase that Mbed TLS's PSA Crypto implementation can be
```
And I have the same suggestion for this commit's message. Too bad we can't make suggestions on commit messages.
(Actually, there are many occurences of PSA below that could be changed to PSA Crypto. Won't list them individually.)

File: tests/crypto/mbedtls_psa/testcase.yaml [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+# supported on any platform. A very minimal configuration is tested here: simply
+# enable support for PSA APIs in Mbed TLS and then test them with:
+# - psa_crypto_init() which is required before any PSA operation
+# - psa_generate_random() which is always available as long as PSA is initialized.
+#
+# Since it might take too long to really execute this test on all Zephyr's
+# supported platforms, we reduce to a very small selected group whose
```
Comment (tomi-font) [inline tests/crypto/mbedtls_psa/testcase.yaml:8]:
```suggestion
# Since it might take too long to execute this test on all platforms
# supported by Zephyr, we reduce to a very small selected group whose
```

File: tests/crypto/mbedtls_psa/testcase.yaml [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+#   latter case will allow to test
+#   CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG.
+common:
+  tags:
+    - mbedtls
+    - psa
+tests:
+  crypto.mbedtls_psa.with_entropy_driver:
+    # Pick a platform which has an entropy driver and enable it.
+    platform_allow:
```
Comment (tomi-font) [inline tests/crypto/mbedtls_psa/testcase.yaml:25]: I think that instead of using `platform_allow` you could use `integration_platforms`. Especially since the behavior of `platform_allow` may be surprising in certain cases (https://github.com/zephyrproject-rtos/zephyr/pull/75962). See https://docs.zephyrproject.org/latest/develop/test/twister.html for reference.

File: tests/crypto/mbedtls_psa/src/main.c [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+/*
+ * Test psa_crypto_init() and psa_generate_random() on the PSA implementation
+ * provided by Mbed TLS (platforms using TFM are filtered out in the yaml file).
+ */
+
+#include <zephyr/ztest.h>
+
+#include "psa/crypto.h"
```
Comment (tomi-font) [inline tests/crypto/mbedtls_psa/src/main.c:14]:
```suggestion
#include <psa/crypto.h>
```

**Review by tomi-font — COMMENTED** [review]
Good stuff!

File: modules/mbedtls/CMakeLists.txt (hunk 8 above)
Comment (valeriosetti) [reply modules/mbedtls/CMakeLists.txt:16]: Yeah, but I saw [somewhere else](https://github.com/zephyrproject-rtos/zephyr/blob/c008cbab1a05316139de191b0553ab6ccc0073ad/subsys/random/CMakeLists.txt#L12) where the "Warning:" part was still added to the message so I thought I could do the same. However I don't have any strong opinion on this, so I can remove that ;)

File: tests/crypto/mbedtls_psa/testcase.yaml (hunk 16 above)
Comment (valeriosetti) [reply tests/crypto/mbedtls_psa/testcase.yaml:25]: Thanks for the tip! I didn't know the difference between `platform_allow` and `integration_platform` before.

File: modules/mbedtls/Kconfig.tls-generic (hunk 6 above)
Comment (valeriosetti) [reply modules/mbedtls/Kconfig.tls-generic:510]: Right, the initial part of the sentence ("the main difference") moves the focus away from security. I rephrased the sentence.

File: modules/mbedtls/zephyr_init.c (hunk 12 above)
Comment (valeriosetti) [reply modules/mbedtls/zephyr_init.c:121]: As for what it concerns with the kconfigs, `CSPRNG_ENABLED` is set whenever `ENTROPY_HAS_DRIVER` [is set](https://github.com/zephyrproject-rtos/zephyr/blob/95622e49b215bcd3ff61f666db6d74eedd4f5b51/subsys/random/Kconfig#L83) and there is no way to disable the `random` subsys so IMO they can be considered as equivalent.

I think that the behavior of `mbedtls_hardware_poll()` is not changed by the new implementation. In the old design the entropy device was polled first (if available) and then `sys_rand_get()` was used as backup. In the new fashion I try `sys_rand_get()` first (again, if available) and then I use `sys_rand_get()` as fallback.
In facts `sys_rand_get()` is implemented by:
- `zephyr/subsys/random/random_ctr_drbg.c` and
- `zephyr/subsys/random/random_entropy_device.c`.
Both implementations get the data from an entropy driver.

In conclusion I think that my implementation basically re-implements the same pattern that we had previously. Am I wrong?

Comment (tomi-font) [reply modules/mbedtls/zephyr_init.c]: I don't think so. I was just checking with you the logic. All good!

File: modules/mbedtls/Kconfig.tls-generic (hunk 9 above)
Comment (valeriosetti) [reply modules/mbedtls/Kconfig.tls-generic:515]: But this would make `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG` as the 3rd option of the choice, whereas my purpose here is to make it as a "feature" of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`.

**Review by tomi-font — APPROVED** [review]

**Review by VivekUppunda — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Comment by tomi-font** [comment]
@valeriosetti Actually, I didn't think of this earlier, but this probably deserves some migration guide entry at least for the renamed Kconfig options.
